### ISSUE TITLE
Clippy should warn against redundant calls to `Option::as_ref` on `Option<&_>`

### ISSUE THREAD

**Issue by JoJoJet** [rust-lang/rust-clippy#11618]
### What it does

`Option::as_ref` is used to convert `&Option<T>` to `Option<&T>`. Clippy should warn against using this function when `T` is already a reference type, unless the destination type requires `Option<&&T>` for some reason.

### Advantage

Code cleanliness and brevity. Sometimes redundant calls to `as_ref` can arise if you refactor a variable from being owned to being passed by reference.

### Example

```rust
fn my_fn(value: Option<&T>) {
    value.as_ref().map(...)
}
```

Could be written as:

```rust
fn my_fn(value: Option<&T>) {
    value.map(...)
}
```

**Comment by Centri3** [comment]
There isn't really any reason to, but you can implement a trait on specifically `Option<&&T>` (without the `T` of course) but not `Option<&T>`, so we'll need to be more careful than just checking whether it's explicitly `Option<&&T>`

Example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f524e3694863e041accfbba1543e8f2a
